cdxev/plausibility_check.py

```diff
@@ -0,0 +1,270 @@
+            for bom_ref in dependency.get("dependsOn", []):
+                if bom_ref not in list_of_actual_bom_refs:
+                    component = get_component_by_ref(
+                        dependency.get("ref", ""), list_of_all_components
+                    )
+                    id = ComponentIdentity.create(component, allow_unsafe=True)
+                    errors.append(
+                        create_error_orphaned_bom_ref(
+                            bom_ref,
+                            "dependencies-dependsOn of reference"
+                            + dependency.get("ref", "")
+                            + f" belonging to component ({id})",
+                        )
+                    )
+
+        else:
+            errors.append(
+                create_error_orphaned_bom_ref(dependency.get("ref", ""), "dependencies")
+            )
+
+    # check compositions
+    for composition in sbom.get("compositions", []):
+        for reference in composition.get("assemblies", []):
+            if reference not in list_of_actual_bom_refs:
+                errors.append(create_error_orphaned_bom_ref(reference, "compositions"))
+    # check vulnearabilities
+    for vulnerability in sbom.get("vulnerabilities", []):
+        for affected in vulnerability.get("affects", []):
+            if affected.get("ref", "") not in list_of_actual_bom_refs:
+                errors.append(
+                    create_error_orphaned_bom_ref(
+                        affected.get("ref", ""),
+                        "vulnerabilitie " + vulnerability.get("id", ""),
+                    )
+                )
+    return errors
+
+
+def check_logic_of_dependencies(sbom: dict) -> list[dict]:
+    """
+    The function checks if the sbom contains circular dependencies,
+    e.g. components, that depend on themself.
+    It also checks if the dependency tree is fully connected
+    (orphaned bom-refs are not considered in this check).
+
+    :param dict sbom: the sbom
+    :return: list with the notifications of found errors
+    :rtype: list[dict]
+    """
+    errors = []
+    list_of_all_components = sbom.get("components", []).copy()
+    list_of_actual_bom_refs = get_bom_refs_from_components(sbom.get("components", []))
+    list_of_actual_bom_refs.append(
+        sbom.get("metadata", {}).get("component", {}).get("bom-ref")
+    )
+    # Check for circular references in dependencies
+    for current_reference in list_of_actual_bom_refs:
+        list_of_upstream_references = get_upstream_dependency_bom_refs(
+            current_reference, sbom.get("dependencies", [])
+        )
+        if current_reference in list_of_upstream_references:
+            errors.append(create_error_circular_reference(current_reference, sbom))
+
+        # check if the dependency tree is connected, i.e. that the product
+        # decribed by the sbom depends directly or indirectly on every component.
+        # also checks that every component is depended on
```

Comment:
This is a misinterpretation of CycloneDX. Not all components in an SBOM are required to be part of the same dependency tree.

That is why *dependencies* and *assemblies* are both a thing in CycloneDX. There are products assembled from multiple components, where there is no *depends-on* relationship but rather an *is-part-of* relationship.